code/__HELPERS/ray_helpers.dm

```diff
@@ -0,0 +1,586 @@
+/**
+  * Any unsorted global procs that help when locating turfs,  via
+  * casting rays out in specified directions etc
+  */
+
+/**
+  * Returns the turf offset from a given atom by the given dx and dy
+  * parameters
+  *
+  * This proc is bound to the map edges so will not overrun them
+  *
+  * Arguments:
+  * * atom/A the atom to calculate the offset from (the source)
+  * * dx, how many steps on the map grid to go in the x direction
+  * * dy, how many steps on the map grid to go in the y direction
+  */
+/proc/get_offset_target_turf(atom/A, dx, dy)
+	var/x = min(world.maxx, max(1, A.x + dx))
+	var/y = min(world.maxy, max(1, A.y + dy))
+	return locate(x,y,A.z)
+
+/**
+  * Get turfs on the edge of a bounding box
+  *
+  * Gets the turfs on the edge of a bounding box of
+  * given x and y height/width that surround the passed
+  * in source atom, uses [/proc/get_offset_target_turf]
+  * internally so will be bounded by map edges
+  *
+  * Arguments:
+  * * atom/A the atom to place the box around
+  * * radius_x radius of the box in the x dimension
+  * * radius_y the radius of the box in the y dimension
+  */
+/proc/get_bounding_turfs(atom/A, radius_x, radius_y)
```

Comment:
what's the difference between using this and `RANGE_TURFS`?
https://github.com/tgstation/tgstation/blob/1aa293ea337283a0191140a878eeba319221e5df/code/__HELPERS/game.dm#L1-L6

Reply:
`get_bounding_turfs` return only border turfs on range.
`RANGE_TURFS` return all turfs in range.

Reply:
```
if(radius_x == radius_y)
	return RANGE_TURFS(radius_x, A) - A
```

Reply:
your result
```
XXXXX
XXXXX
XX0XX
XXXXX
XXXXX
```
get_bounding_turfs result
```
XXXXX
X000X
X000X
X000X
XXXXX
```

Reply:
```cpp
#define BOUNDING_TURFS(radius, center) (RANGE_TURFS(radius, center) - RANGE_TURFS(radius - 1, center))
```

Reply:
proof or get out

Reply:
Okay, I was wronger than wrong. My method is not just inefficient, it's trash. The `list - list` operation is incredibly slow and makes the proc over a hundred times more expensive. Good to note.

So after finding out that I played with things a bit more. Firstly, this proc you made doesn't return the right values. I tweaked it to remove costly `|=` operations and redundant loops and ended up with this:
```cpp
/proc/get_bounding_turfs(atom/A, r_x, r_y)
 var/start_x = A.x - r_x
 var/end_x = A.x + r_x
 var/start_y = A.y - r_y
 var/end_y = A.y + r_y
 . = list()
 for(var/y in start_y to end_y)
  . += O_T_T(A, start_x, y)
  . += O_T_T(A, end_x, y)
 for(var/x in start_x + 1 to end_x - 1)
  . += O_T_T(A, x, start_y)
  . += O_T_T(A, x, end_y)
```
It returns the intended values and should be a bit cheaper than the current version. But then I re-did the method using block:
```cpp
//SE to NE + SE to SW w/o corners + NE to NW w/o corners + SW to NW
#define BOUNDING_TURFS(center, radius_x, radius_y) block(\
 OFFSET_TARGET_TURF(center, -radius_x, -radius_y),\
 OFFSET_TARGET_TURF(center, -radius_x, radius_y)\
 ) + block(\
 OFFSET_TARGET_TURF(center, -radius_x + 1, -radius_y),\
 OFFSET_TARGET_TURF(center, radius_x - 1, -radius_y)\
 ) + block(\
 OFFSET_TARGET_TURF(center, -radius_x + 1, radius_y),\
 OFFSET_TARGET_TURF(center, radius_x - 1, radius_y)\
 ) + block(\
 OFFSET_TARGET_TURF(center, radius_x, -radius_y),\
 OFFSET_TARGET_TURF(center, radius_x, radius_y)\
 )
```
Note: I was using the offset target turf define on both to reduce overhead:
```cpp
#define OFFSET_TARGET_TURF(source, dx, dy) ( locate( clamp(source.x + dx, 1, world.maxx), clamp(source.y + dy, 1, world.maxy), source.z ) )
```
The result of my profiling:
```
BOUNDING_TURFS in 2 sec
get_bounding_turfs in 16.1 sec
```

Reply:
If you just ctrl+c and ctrl+v `get_bounding_turfs` for `BOUNDING_TURFS` around the code, things should work the same as you intended before, as the input parameters are the same.